### PR TITLE
chore: Publish container image on release

### LINKED ISSUE #220: feat: Build docker image for each tag and save them in repo
Relates #217

### PR DESCRIPTION
**Pull request by MindTooth**
Put an `x` into the box if that apply:

- [ ] This PR introduces breaking change.
- [ ] This PR fixes a bug.
- [x] This PR adds new functionality.
- [ ] This PR enhances existing functionality.

### Description of your changes

A basic workflow for building the image and publishing it to the registry of this project.  Builds on a new release and pushes to `master`.

Fixes #220.

### How has this code been tested

I've built and pushed images to my own fork.

### CODE REVIEW

File: .github/workflows/build-image.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+name: publish container image
+
+on:
+  push:
```
Comment (antonbabenko) [inline .github/workflows/build-image.yaml:5]: Could you also add `workflow_dispatch` to be able to trigger this action manually (without pushing to `master`)? See this - https://github.com/terraform-aws-modules/terraform-aws-rds-aurora/blob/master/.github/workflows/release.yml#L4

File: .github/workflows/build-image.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Set tag for image
+        run: |
+          echo IMAGE_TAG=$([ ${{ github.ref_type }} == 'tag' ] && echo ${{ github.ref_name }} || echo 'latest') >> $GITHUB_ENV
+      - name: Build and Push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          platforms: linux/amd64
```
Comment (MindTooth) [inline .github/workflows/build-image.yaml:36]: I removed `linux/arm64` from this line.  Seeing as the `Dockerfile` only download binaries for x64 I don't see the need for building for more architectures.

**Review by yermulnik — APPROVED** [review]
LGTM (though I'm not familiar with GH actions yet)

File: .github/workflows/build-image.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+---
+name: publish container image
+
+on:
+  push:
+    branches:
+      - "master"
```
Comment (antonbabenko) [inline .github/workflows/build-image.yaml:7]:
```suggestion
      - master
```

File: .github/workflows/build-image.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        run: |
+          echo IMAGE_TAG=$([ ${{ github.ref_type }} == 'tag' ] && echo ${{ github.ref_name }} || echo 'latest') >> $GITHUB_ENV
+      - name: Build and Push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          platforms: linux/amd64
+          push: true
+          tags: |
+            ghcr.io/${{ env.IMAGE_REPOSITORY }}:${{ env.IMAGE_TAG }}
```
Comment (antonbabenko) [inline .github/workflows/build-image.yaml:39]: `IMAGE_REPOSITORY` should be hardcoded or how will it be determined?

Comment (MindTooth) [reply .github/workflows/build-image.yaml:39]: Sorry!  Nice catch.  I had to tweak this as I have an upper case in my username.  Give me 2 sec.

File: .github/workflows/build-image.yaml (hunk 2 above)
Comment (MaxymVlasov) [reply .github/workflows/build-image.yaml:36]: Mac M1 users can run amd64 docker-images in compatible mode, so you're right.

File: .github/workflows/build-image.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        uses: docker/setup-buildx-action@v1
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Set tag for image
+        run: |
+          echo IMAGE_TAG=$([ ${{ github.ref_type }} == 'tag' ] && echo ${{ github.ref_name }} || echo 'latest') >> $GITHUB_ENV
```
Comment (MaxymVlasov) [inline .github/workflows/build-image.yaml:31]: Not sure that we'd like tag image as `latest` when manually triggered.

```suggestion
          echo IMAGE_TAG=$([ ${{ github.ref_type }} == 'tag' ] && echo ${{ github.ref_name }} || echo ${{ github.sha }} ) >> $GITHUB_ENV
```

File: .github/workflows/build-image.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+---
+name: publish container image
+
+on:
+  push:
+    branches:
+      - master
```
Comment (MaxymVlasov) [inline .github/workflows/build-image.yaml:7]: https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#onpushpull_requestbranchestags
```suggestion
   tags:        
      - '**'  
```

File: .github/workflows/build-image.yaml (hunk 5 above)
Comment (MaxymVlasov) [reply .github/workflows/build-image.yaml:31]: @antonbabenko what you think?

**Comment by MaxymVlasov** [comment]
Ignore "Common issues check / pre-commit (pull_request)" fail, it has a bug, trying to merge changes from the master branch, but it should not do this. I will fix that soon.

File: .github/workflows/build-image.yaml (hunk 5 above)
Comment (antonbabenko) [reply .github/workflows/build-image.yaml:31]: `master` means `latest`. This is expected behavior I would say. `github.sha` is not so useful in this context.

File: .github/workflows/build-image.yaml (hunk 6 above)
Comment (MindTooth) [reply .github/workflows/build-image.yaml:7]: If so, we must remove release and master (latest) as it won’t build.  Tags are usually release as well.

**Comment by MindTooth** [comment]
How are we on this?  I’ll happily change anything warranted.

**Comment by antonbabenko** [comment]
@MindTooth It looks good. We just need to get GH Actions green again.

@MaxymVlasov GH Actions fails rightly because it can't push changes back to the fork:
```
remote: Permission to MindTooth/pre-commit-terraform.git denied to github-actions[bot].
fatal: unable to access 'MindTooth/pre-commit-terraform.git': The requested URL returned error: 403
```

Could you please update it, so that we can merge this one?

**Comment by MindTooth** [comment]
Rebased on `master`.  Seems green now.

**Comment by antonbabenko** [comment]
Thanks, @MindTooth ! 🎉 

[v1.58.0](https://github.com/antonbabenko/pre-commit-terraform/releases/v1.58.0) has been just released.

https://github.com/antonbabenko/pre-commit-terraform/pkgs/container/pre-commit-terraform - looks good, too.

**Comment by MaxymVlasov** [comment]
@MindTooth just merge changes from master branch, it includes GHA fix

**Comment by MindTooth** [comment]
You're welcome. :smile:

However, I just noticed that none of the tools got bundled.  Maybe we should enable them all?  Thoughts?

That was the real reason why I submitted this pull request so that I had an image that was "complete". 😔

**Comment by antonbabenko** [comment]
LOL. I hope @MaxymVlasov will be able to fix it :)

**Comment by MaxymVlasov** [comment]
Hah

Need to add `--build-arg INSTALL_ALL=true`

Also, in readme need provide `docker pull instruction` and change image name to pre-commit-terraform to unify name of GH image and local build

**Comment by MaxymVlasov** [comment]
Feel free to send PR or I will do it by myself in ~Tue

**Pull request merged**